### DISCUSSION TITLE
How to handle a case where a mutation can return either a result or error?

### DISCUSSION THREAD

**Discussion by todor-a** [TanStack/query#8066] [Q&A] [answered]
Hey. I'm working with an API that is typed in a way that it can return either a result or an error. What I struggle with is figuring out how to call `onSuccess` only if the API call returns a result. The errors will be returned with a 400 status code, which will invoke the `onError` handler, but I still can't figure out how to return the result, without using `!`.

Any ideas? Here is a minimal [example](https://tsplay.dev/wRQJxW):
```ts
import { useMutation } from '@tanstack/react-query';

declare function get(): Promise<Result>;

type Result = {
    id: string;
    error?: never;
} | {
    id?: never;
    error: number
}

export const useCreate = () => {
    return useMutation<string, number>({
        mutationFn: async () => {
            const result = await get();

            return result.id;
            // ^ type error since this can be undefined 
        },
        onSuccess: (result) => {
            //  how to call this only if the result's error is undefined
        }
    });
};
```

**Comment by TkDodo** [comment, marked as answer]
you need to return a rejected promise (or throw an error) in the `mutationFn`

**Reply by todor-a** [reply]
Thank!
